File: python/time_map.py
class TimeMap(object):

    def __init__(self):
        self.times = {}        

    def set(self, key, value, timestamp):
        """
        :type key: str
        :type value: str
        :type timestamp: int
        :rtype: None
        """
        if key not in self.times:
            self.times[key] = []
        self.times[key].append([value, timestamp])
        

    def get(self, key, timestamp):
        """
        :type key: str
        :type timestamp: int
        :rtype: str
        """
        if key not in self.times:
            return ""
        pairs = self.times[key]
        l = 0
        r = len(pairs) - 1
        value = ""
        while l <= r:
            m = l + (r - l) // 2
            if pairs[m][1] == timestamp:
                return pairs[m][0]
            if pairs[m][1] < timestamp:
                value = pairs[m][0]
                l = m + 1
            else:
                r = m - 1
        return value
        

File: src/main/java/june_2025/TimeMap.java
package june_2025;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TimeMap {

    Map<String, List<Pair>> times;

    public TimeMap() {
        times = new HashMap<>();
    }

    public void set(String key, String value, int timestamp) {
        times.computeIfAbsent(key, k -> new ArrayList<>()).add(new Pair(value, timestamp));
    }

    public String get(String key, int timestamp) {
        if (!times.containsKey(key)) {
            return "";
        }
        List<Pair> pairs = times.get(key);

        int l = 0;
        int r = pairs.size() - 1;
        String value = "";
        while (l <= r) {
            int m = l + (r - l) / 2;
            Pair pair = pairs.get(m);
            if (pair.timestamp == timestamp) {
                return pair.value;
            }
            if (pair.timestamp < timestamp) {
                value = pair.value;
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
        return value;
    }


    static class Pair {
        String value;
        int timestamp;

        public Pair(String value, int timestamp) {
            this.value = value;
            this.timestamp = timestamp;
        }
    }
}
